cupy/_core/_kernel.pyx

```diff
@@ -893,13 +904,32 @@ cdef function.Function _get_ufunc_kernel(
                     ' ? 1 : 0' if arginfo.dtype == bool_ and x != bool_ else ''
                 ))
 
+    out_op = []
     for i, x in enumerate(out_types):
         arginfo = arginfos[i + len(in_types)]
-        types.append(('out%d_type' % i, arginfo.dtype))
-        op.append('out{0}_type &out{0} = _raw_out{0}[_ind.get()];'.format(i))
+        types.append(('out%d_type' % i, x))
+        if arginfo.dtype == x:
+            op.append(
+                'out{0}_type &out{0} = _raw_out{0}[_ind.get()];'.format(i))
+        else:
+            op.append(
+                'out{0}_type out{0}({1});'
+                .format(
+                    i,
+                    fix_cast_expr(arginfo.dtype, x, f'_raw_out{i}[_ind.get()]')
+                ))
+            out_op.append(
+                '_raw_out{0}[_ind.get()] = {1};'.format(
+                    i,
+                    fix_cast_expr(x, arginfo.dtype, f'out{i}')
+                )
+            )
```

Comment:
This "load and save" approach is not really a good idea.
1. The load is not necessary.  I'd remove the ability of reading `out` from `cupy.ufunc`, while it's fine with `cupy.ElementwiseKernel`.
2. The save causes "precision loss" if no-op is expected.  For example, CuPy should carefully implement `where` of ufunc.

The current problem is just a special case of 2.: `cupy.copyto(..., where=...)` is getting wrong with
```python
elementwise_copy_where = create_ufunc(
    'cupy_copy_where',
    ('??->?', 'b?->b', 'B?->B', 'h?->h', 'H?->H', 'i?->i', 'I?->I', 'l?->l',
     'L?->L', 'q?->q', 'Q?->Q', 'e?->e', 'f?->f', 'd?->d', 'F?->F', 'D?->D'),
    'if (in1) out0 = in0',
    default_casting='unsafe')
```
, which was found in the tests of `cupy.unwrap`.  (BTW,  for float dtypes, `cupy.testing.shaped_arange` could use `arange(size) + 0.5` instead of `+ 1` in order to detect this kind of bugs.)
https://github.com/cupy/cupy/blob/f06d85b3127b46a10346f79e6e6c351799580a27/cupy/_math/trigonometric.py#L153-L154

To resolve it, either
- implement (private) `_where` arg of `ufunc.__call__` that works at least for `cupy_copy`, or
- rewrite the `cupy_copy_where` kernel as `cupy.ElementwiseKernel`.